### PR TITLE
1

### PR DESCRIPTION
**Pull request by tongkiat**
### Solutions for Chapter {1}

cc @vrom911 @chshersh: All comments, feedback, and suggestions are greatly appreciated.  Thank you.

### CODE REVIEW

File: src/Chapter1.hs [hunk 1]
```diff
@@ -70,8 +68,12 @@ Each Haskell module starts with the "module <MODULE_NAME> where" line.
 Modules should have the same name as the corresponding file with
 the `.hs` extension.
 -}
+
 module Chapter1 where
 
+import qualified Data.Char
+import qualified Data.List
```
Comment (vrom911) [inline src/Chapter1.hs:75]: Intriguing start! 🤩

File: src/Chapter1.hs [hunk 2]
```diff
@@ -415,18 +427,20 @@ task is to specify the type of this function.
 
 >>> squareSum 3 4
 49
+
 -}
 
+squareSum :: Num a => a -> a -> a
```
Comment (vrom911) [inline src/Chapter1.hs:433]: We explain `Num` in the following chapters, so you are one step ahead 😸

In here `Int -> Int -> Int` would also work, as a more specific type 🙂

File: src/Chapter1.hs [hunk 3]
```diff
@@ -436,8 +450,9 @@ Implement the function that takes an integer value and returns the next 'Int'.
   every type ｡.☆.*｡. No need to worry much about "error" here, just replace the
   function body with the proper implementation.
 -}
-next :: Int -> Int
-next x = error "next: not implemented!"
+
+next :: Enum a => a -> a
+next = succ
```
Comment (vrom911) [inline src/Chapter1.hs:455]: 👍🏼 `Enum`! You rock 🎸

File: src/Chapter1.hs [hunk 4]
```diff
@@ -477,9 +492,9 @@ Implement a function that returns the last digit of a given number.
   results. Or you can try to guess the function name, search for it and check
   whether it works for you!
 -}
--- DON'T FORGET TO SPECIFY THE TYPE IN HERE
-lastDigit n = error "lastDigit: Not implemented!"
 
+lastDigit :: Integral a => a -> a
+lastDigit n = (abs n) `mod` 10
```
Comment (vrom911) [inline src/Chapter1.hs]: Nice work on using infix form of calling functions! We don't explain this syntax in our course to make it short. But well done :+1: 

Also, you can remove `()`, they are redundant in this case :slightly_smiling_face: 
```suggestion
lastDigit n = abs n `mod` 10
```

File: src/Chapter1.hs [hunk 5]
```diff
@@ -507,9 +522,9 @@ branches because it is an expression and it must always return some value.
 👩‍🔬 Due to lazy evaluation in Haskell, only the expression from the branch
   satisfying the check will be returned and, therefore, evaluated.
 -}
-closestToZero :: Int -> Int -> Int
-closestToZero x y = error "closestToZero: not implemented!"
 
+closestToZero :: (Ord p, Num p) => p -> p -> p
```
Comment (vrom911) [inline src/Chapter1.hs:526]: I think this function makes more sense for concrete types instead:

```suggestion
closestToZero :: Int -> Int -> Int
```

File: src/Chapter1.hs [hunk 6]
```diff
@@ -553,11 +569,16 @@ Implement a function that checks whether a given character is a vowel.
...
 >>> isVowel 'x'
 False
+
 -}
-isVowel c = error "isVowel: not implemented!"
 
+isVowel :: Char -> Bool
+isVowel char = any (c ==) "aeiou"
+  where
+    c = Data.Char.toLower char
```
Comment (vrom911) [inline src/Chapter1.hs:581]: Nice catch! 👌🏼

File: src/Chapter1.hs [hunk 7]
```diff
@@ -553,11 +569,16 @@ Implement a function that checks whether a given character is a vowel.
...
 True
+
 >>> isVowel 'x'
 False
+
 -}
-isVowel c = error "isVowel: not implemented!"
 
+isVowel :: Char -> Bool
+isVowel char = any (c ==) "aeiou"
```
Comment (vrom911) [inline src/Chapter1.hs]: This is just `elem` function 🙂 

```suggestion
isVowel char = c `elem` "aeiou"
```

File: src/Chapter1.hs [hunk 8]
```diff
@@ -634,15 +661,19 @@ Implement a function that returns the first digit of a given number.
...
 aren't ready for this boss yet!
 -}
 
-firstDigit n = error "firstDigit: Not implemented!"
-
+firstDigit :: Integral a => a -> a
+firstDigit n = go (abs n `divMod` 10)
+  where
+    go (0, m) = m
+    go (d, _) = go (d `divMod` 10)
```
Comment (vrom911) [inline src/Chapter1.hs:676]: Very creative solution! 👍🏼

**Review by vrom911 — APPROVED** [review]
Nice work! 👍🏼 Feel free to merge this one :)

**Comment by tongkiat** [comment]
Thank you, @vrom911 ! :smile:

**Pull request merged**